src/main/drivers/serial_pinconfig.c

```diff
@@ -0,0 +1,275 @@
+# endif
+# if !defined(INVERTER_PIN_UART6)
+#  define INVERTER_PIN_UART6 NONE
+# endif
+#endif
+
+#ifdef USE_UART7
+# if !defined(UART7_RX_PIN)
+#  define UART7_RX_PIN NONE
+# endif
+# if !defined(UART7_TX_PIN)
+#  define UART7_TX_PIN NONE
+# endif
+# if !defined(INVERTER_PIN_UART7)
+#  define INVERTER_PIN_UART7 NONE
+# endif
+#endif
+
+#ifdef USE_UART8
+# if !defined(UART8_RX_PIN)
+#  define UART8_RX_PIN NONE
+# endif
+# if !defined(UART8_TX_PIN)
+#  define UART8_TX_PIN NONE
+# endif
+# if !defined(INVERTER_PIN_UART8)
+#  define INVERTER_PIN_UART8 NONE
+# endif
+#endif
+
+#ifdef USE_SOFTSERIAL1
+# if !defined(SOFTSERIAL1_RX_PIN)
+#  define SOFTSERIAL1_RX_PIN NONE
+# endif
+# if !defined(SOFTSERIAL1_TX_PIN)
+#  define SOFTSERIAL1_TX_PIN NONE
+# endif
+#endif
+
+#ifdef USE_SOFTSERIAL2
+# if !defined(SOFTSERIAL2_RX_PIN)
+#  define SOFTSERIAL2_RX_PIN NONE
+# endif
+# if !defined(SOFTSERIAL2_TX_PIN)
+#  define SOFTSERIAL2_TX_PIN NONE
+# endif
+#endif
+
+// XXX Should these be moved to drivers/serial.c?
+
+#if defined(USE_UART) || defined(USE_SOFTSERIAL1) || defined(USE_SOFTSERIAL1)
+typedef struct serialDefaultPin_s {
+    serialPortIdentifier_e ident;
+    ioTag_t rxIO, txIO, inverterIO;
+} serialDefaultPin_t;
+
+static const serialDefaultPin_t serialDefaultPin[] = {
+#ifdef USE_UART1
+    { SERIAL_PORT_USART1, DEFIO_TAG_E(UART1_RX_PIN), DEFIO_TAG_E(UART1_TX_PIN), DEFIO_TAG_E(INVERTER_PIN_UART1) },
+#endif
+#ifdef USE_UART2
+    { SERIAL_PORT_USART2, DEFIO_TAG_E(UART2_RX_PIN), DEFIO_TAG_E(UART2_TX_PIN), DEFIO_TAG_E(INVERTER_PIN_UART2) },
+#endif
+#ifdef USE_UART3
+    { SERIAL_PORT_USART3, DEFIO_TAG_E(UART3_RX_PIN), DEFIO_TAG_E(UART3_TX_PIN), DEFIO_TAG_E(INVERTER_PIN_UART3) },
+#endif
+#ifdef USE_UART4
+    { SERIAL_PORT_UART4, DEFIO_TAG_E(UART4_RX_PIN), DEFIO_TAG_E(UART4_TX_PIN), DEFIO_TAG_E(INVERTER_PIN_UART4) },
+#endif
+#ifdef USE_UART5
+    { SERIAL_PORT_UART5, DEFIO_TAG_E(UART5_RX_PIN), DEFIO_TAG_E(UART5_TX_PIN), DEFIO_TAG_E(INVERTER_PIN_UART5) },
+#endif
+#ifdef USE_UART6
+    { SERIAL_PORT_USART6, DEFIO_TAG_E(UART6_RX_PIN), DEFIO_TAG_E(UART6_TX_PIN), DEFIO_TAG_E(INVERTER_PIN_UART6) },
+#endif
+#ifdef USE_UART7
+    { SERIAL_PORT_USART7, DEFIO_TAG_E(UART7_RX_PIN), DEFIO_TAG_E(UART7_TX_PIN), DEFIO_TAG_E(INVERTER_PIN_UART7) },
+#endif
+#ifdef USE_UART8
+    { SERIAL_PORT_USART8, DEFIO_TAG_E(UART8_RX_PIN), DEFIO_TAG_E(UART8_TX_PIN), DEFIO_TAG_E(INVERTER_PIN_UART8) },
+#endif
+#ifdef USE_SOFTSERIAL1
+    { SERIAL_PORT_SOFTSERIAL1, DEFIO_TAG_E(SOFTSERIAL1_RX_PIN), DEFIO_TAG_E(SOFTSERIAL1_TX_PIN), DEFIO_TAG_E(NONE) },
+#endif
+#ifdef USE_SOFTSERIAL2
+    { SERIAL_PORT_SOFTSERIAL2, DEFIO_TAG_E(SOFTSERIAL2_RX_PIN), DEFIO_TAG_E(SOFTSERIAL2_TX_PIN), DEFIO_TAG_E(NONE) },
+#endif
+};
+
+PG_REGISTER_WITH_RESET_FN(serialPinConfig_t, serialPinConfig, PG_SERIAL_PIN_CONFIG, 0);
+
+void pgResetFn_serialPinConfig(serialPinConfig_t *serialPinConfig)
+{
+    memset(serialPinConfig, 0, sizeof(serialPinConfig_t) * SERIAL_PORT_MAX_INDEX);
```

Comment:
Well, yes ... array vs pointer do complicate this, I didn't think about this before.

If array is passed to function (as pointer), you need to specify array size explicitly. `memset(ptr, 0, len * sizeof(*ptr));`  would be my prefered way.

When only structure is passed, it's equivalent to single-element array, so ` memset(ptr, 0, sizeof(*ptr));`

With array typed variable, best option is probably `memset(&array, 0, sizeof(array));` - it will clear variable of any type (setting pointer value to null/ working correctly for arrays)

Here is excelent explanation of array to pointer degradation : http://stackoverflow.com/questions/17506138/when-is-an-array-name-or-a-function-name-converted-into-a-pointer-in-c/17506236#17506236


In all cases I'll try to avoid using variable type explicitly. The code will be perfectly legal when variable type is changed, possibly using to ugly errors ...